x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/relevance/RelevanceSettingsService.java

```diff
@@ -21,16 +40,21 @@ public class RelevanceSettingsService {
     public static final String RELEVANCE_SETTINGS_PREFIX = "relevance_settings-";
     private final Client client;
 
+    private static final Logger logger = LogManager.getLogger(RelevanceSettingsService.class);
+
     public RelevanceSettingsService(final Client client) {
         this.client = client;
     }
 
     public RelevanceSettings getRelevanceSettings(String settingsId) throws RelevanceSettingsNotFoundException,
         RelevanceSettingsInvalidException {
         // TODO cache relevance settings, including cache invalidation
-        final Map<String, Object> settingsContent = client.prepareGet(ENT_SEARCH_INDEX, RELEVANCE_SETTINGS_PREFIX + settingsId)
-            .get()
-            .getSource();
+        Map<String, Object> settingsContent = null;
+        try {
+            settingsContent = client.prepareGet(ENT_SEARCH_INDEX, RELEVANCE_SETTINGS_PREFIX + settingsId).get().getSource();
+        } catch (IndexNotFoundException e) {
+            ensureInternalIndex(client);
```

Comment:
Nice! So the index will be created when a query with specified relevance settings is done. 

Would there be any other way of ensuring this index exist, like maybe at service creation (constructor)? That would ensure users are able to add settings to it before doing an actual search with them...

Reply:
adding it to the constructor does not work:

```
[2022-10-10T13:56:52,889][INFO ][o.e.x.r.r.RelevanceSettingsService] [runTask-0] Failed to create .ent-search index java.lang.IllegalStateException: NodeClient has not been initialized
```

I add it as part of the query builder - let me know if that works.
This means the index will first be created when `relevance_match` is used for the first time.

Reply:
what I think I can do instead is to have `RelevanceSettingsService` implement `ClusterStateListener` so that it actually tries to create the internal index when the cluster is ready 🤔